### ISSUE TITLE
Wave SVG flags

### ISSUE THREAD

**Issue by behdad** [googlefonts/noto-emoji#3]
We want to explore options for generating waved flag SVGs. Detailed discussion can be found in the original thread.

Moved from googlei18n/noto-fonts#358

**Comment by behdad** [comment]
cc @behdad @roozbehp @dougfelt @jamesgk @raphlinus

**Comment by asturur** [comment]
this problem is the reason why there are not svg flags available?

**Comment by dougfelt** [comment]
That, plus we don't currently use SVG in fonts on Android or CrOS so we haven't pushed on this ourselves.

That said, the svg flag sources are available in third-party/region-flags/svg.

**Comment by behdad** [comment]
My current thinking is to forgo any fancy operators or mesh gradients and just read the SVG, transform path data, and save it.  Things like gradients, etc, become tricky though.  We have to see how bad those become in practice.

**Comment by dougfelt** [comment]
Transforming the path data seems  like it would result in fairly complex paths.  The mesh gradients seemed like an interesting approach, since if I remember correctly it is applied post-rasterization.  Also seemed like they could be implemented on graphics hardware and so fairly low-cost.  But if there's no implementation there's no implementation...

**Comment by asturur** [comment]
you want to generate wavy SVG from rectangular SVG flags?

or wavy PNGs from rectangular SVG flags?

**Comment by dougfelt** [comment]
Neither.  Behdad's code generates wavy PNGs from mostly rectangular PNGs now, by rasterizing the PNG, moving pixels around, and creating a new PNG.  Similarly, we want to generate wavy SVGs from mostly rectangular SVGs, by applying some kind of transform to the SVG.  Behdad's current proposal is to modify/create new outlines/operations, it seems.  A few years ago there were (proposed?) SVG operators that (again, if I recall correctly) apply a transform to the component they control.  The transform is effectively applied as a late step in SVG rasterization, and modifies the pixels in the raster after the graphic has been stroked/filled etc.  Using this approach we'd wrap the flag graphic with these SVG operations, but it requires support for them in the SVG rasterizer.  But if there's no support, we can't do that.

**Comment by behdad** [comment]
Full discussion is in https://github.com/googlei18n/noto-fonts/issues/358
Even if feDisplacementMap is available, mesh gradients are a SVG 2 thing, which means they may never arrive...  Another issue with that approach is that unless we do supersampling (like we do, 8x, in waveflag.c), the results look really bad.  So, looks like that approach is for sure dead.

The transform-paths approach is not that bad.  At the simplest we can just sample all lines and curves into tiny line segments, transform all the end points, then feed this through the same path simplification filter that we need for simplifying the flags anyway.

**Comment by behdad** [comment]
> you want to generate wavy SVG from rectangular SVG flags?

Yes.  Not sure why Doug said neither.

**Comment by asturur** [comment]
some flags have some rich image drawing that has been trough some kind of raster to vector approach. how is that thing going to be transformed? just curiosity.

**Comment by behdad** [comment]
> some flags have some rich image drawing that has been trough some kind of raster to vector approach. how is that thing going to be transformed? just curiosity.

We just need to implement transforming a Bezier over a Bezier mesh.  As I said: simplest is to just convert Bezier to a line spline, transform the points, and reconstruct a Bezier spline from it.

**Comment by rsheeter** [comment]
@anthrotype I fear we will have to deal with this for COLR

**Comment by anthrotype** [comment]
after #389, waved vector region-flags are now in https://github.com/googlefonts/noto-emoji/tree/main/third_party/region-flags/waved-svg

These were produced using the https://github.com/rsheeter/warp tool (specifically, the wave_ninja.py script in that repository, with --noto_dir flag pointing to the noto-emoji's third_party/region-flags/svg folder).

We can finally close this old issue I guess?

**Comment by behdad** [comment]
Lovely!
